src/core/consumerthread.h

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_CORE_CONSUMERTHREAD_H
+#define BITCOIN_CORE_CONSUMERTHREAD_H
+
+#include <future>
+#include <thread>
+
+#include <core/producerconsumerqueue.h>
+#include <util.h>
+
+template <WorkerMode MODE>
+class ConsumerThread;
+
+//! A WorkItem() encapsulates a task that can be processed by a ConsumerThread()
+//! @see ConsumerThread()
+template <WorkerMode MODE>
+class WorkItem
+{
+    friend ConsumerThread<MODE>; //<! only a consumer thread can execute a WorkItem
+
+protected:
+    WorkItem(){};
+    virtual void operator()(){};
+};
+
+template <WorkerMode MODE>
+class GenericWorkItem : public WorkItem<MODE>
+{
+    friend ConsumerThread<MODE>;
+
+public:
+    GenericWorkItem(std::function<void()> f) : m_f(f) {}
+
+protected:
+    void operator()() override
+    {
+        m_f();
+    }
+
+    std::function<void()> m_f;
+};
+
+//! A special WorkItem() that is used to interrupt a blocked ConsumerThread() so that it can terminate
+template <WorkerMode MODE>
+class ShutdownPill : public WorkItem<MODE>
```

Comment:
In commit "Add ConsumerThread: to consumer and operate on work from a ProducerConsumerQueue" (7f8a8889564b6509285fbab64825fecb983fca6d)

I think I agree with James. Even if you want to support shutting down specific threads, an approach more like "Everybody wake up, check if you are supposed to exit, and if not go back to sleep again" might be simpler than "Random thread wake up, check if you are supposed to exit, and if not wake up another random thread, but not if you already seen this particular notification before," or whatever the correct description is. (If you do want to stick with the current approach, it could be nice to add a high level comment explaining it like this.)